File: sdk/c/src/providers/fake_email_site.c
/**
 * FakeEmailSite -- https://fake-email.site
 */

#include "tempmail_internal.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define FAKE_EMAIL_SITE_BASE "https://api.fake-email.site"

static const char *fake_email_site_headers[] = {
    "Accept: application/json",
    "Content-Type: application/json",
    NULL
};

/**
 * 创建临时邮箱
 * POST /api/temporary-address body: {}
 * 返回 {"temp_email_addr":"xxx@fake-email.site"}
 */
tm_email_info_t* tm_provider_fake_email_site_generate(void) {
    tm_http_response_t *resp = tm_http_request(
        TM_HTTP_POST,
        FAKE_EMAIL_SITE_BASE "/api/temporary-address",
        fake_email_site_headers,
        "{}",
        15
    );
    if (!resp || resp->status < 200 || resp->status >= 300) {
        tm_http_response_free(resp);
        return NULL;
    }

    cJSON *root = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!root) return NULL;

    const char *email = TM_JSON_STR(cJSON_GetObjectItemCaseSensitive(root, "temp_email_addr"), "");
    if (!email[0] || !strchr(email, '@')) {
        cJSON_Delete(root);
        return NULL;
    }

    tm_email_info_t *info = tm_email_info_new();
    if (!info) {
        cJSON_Delete(root);
        return NULL;
    }
    info->channel = CHANNEL_FAKE_EMAIL_SITE;
    info->email = tm_strdup(email);
    info->token = tm_strdup(email);
    cJSON_Delete(root);
    return info;
}

/**
 * 对字符串进行 URL 编码
 */
static char *fake_email_site_urlencode(const char *s) {
    if (!s) return tm_strdup("");
    size_t len = strlen(s);
    char *out = (char*)malloc(len * 3 + 1);
    if (!out) return NULL;
    char *p = out;
    static const char hex[] = "0123456789ABCDEF";
    for (size_t i = 0; i < len; i++) {
        unsigned char c = (unsigned char)s[i];
        if ((c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') ||
            (c >= '0' && c <= '9') || c == '-' || c == '_' || c == '.' || c == '~' || c == '@') {
            *p++ = (char)c;
        } else {
            *p++ = '%';
            *p++ = hex[c >> 4];
            *p++ = hex[c & 15];
        }
    }
    *p = '\0';
    return out;
}

/**
 * 获取邮件列表
 * GET /api/inbox/poll?address=xxx
 * 返回 {"temp_email_addr":"xxx","new_mail_count":0,"next_since":null,"messages":[]}
 * 消息字段: id, from_addr, to_addr, subject, body_text, received_at
 * 只有 body_text 没有原生 html，tm_normalize_email 会自动补 html
 * 处理 404 返回 NULL 且 count=0
 */
tm_email_t* tm_provider_fake_email_site_get_emails(const char *email, int *count) {
    *count = -1;
    if (!email || !email[0]) return NULL;

    char *enc = fake_email_site_urlencode(email);
    if (!enc) return NULL;
    size_t cap = strlen(FAKE_EMAIL_SITE_BASE) + strlen(enc) + 64;
    char *url = (char*)malloc(cap);
    if (!url) {
        free(enc);
        return NULL;
    }
    snprintf(url, cap, "%s/api/inbox/poll?address=%s", FAKE_EMAIL_SITE_BASE, enc);
    free(enc);

    tm_http_response_t *resp = tm_http_request(TM_HTTP_GET, url, fake_email_site_headers, NULL, 15);
    free(url);
    if (!resp) return NULL;

    /* 处理 404 返回空 */
    if (resp->status == 404) {
        tm_http_response_free(resp);
        *count = 0;
        return NULL;
    }

    if (resp->status < 200 || resp->status >= 300) {
        tm_http_response_free(resp);
        return NULL;
    }

    cJSON *root = cJSON_Parse(resp->body);
    tm_http_response_free(resp);
    if (!root) return NULL;

    cJSON *arr = cJSON_GetObjectItemCaseSensitive(root, "messages");
    int n = (arr && cJSON_IsArray(arr)) ? cJSON_GetArraySize(arr) : 0;
    *count = n;
    if (n == 0) {
        cJSON_Delete(root);
        return NULL;
    }

    tm_email_t *emails = tm_emails_new(n);
    if (!emails) {
        *count = -1;
        cJSON_Delete(root);
        return NULL;
    }

    for (int i = 0; i < n; i++) {
        cJSON *raw = cJSON_GetArrayItem(arr, i);
        emails[i] = tm_normalize_email(raw, email);
    }
    cJSON_Delete(root);
    return emails;
}

File: sdk/rust/src/providers/fake_email_site.rs
/*!
 * FakeEmailSite — https://fake-email.site
 */

use crate::config::{block_on, http_client};
use crate::normalize::normalize_email;
use crate::types::{Channel, Email, EmailInfo};
use serde_json::{json, Value};

const API_BASE: &str = "https://api.fake-email.site/api";

fn headers(b: wreq::RequestBuilder) -> wreq::RequestBuilder {
    b.header("Accept", "application/json")
        .header("Content-Type", "application/json")
}

/// 创建临时邮箱
pub fn generate_email() -> Result<EmailInfo, String> {
    block_on(async {
        let resp = headers(http_client().post(format!("{}/temporary-address", API_BASE)))
            .body(json!({}).to_string())
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if !resp.status().is_success() {
            return Err(format!("fake-email-site create {}", resp.status()));
        }
        let data: Value = resp.json().await.map_err(|e| e.to_string())?;
        let email = data
            .get("temp_email_addr")
            .and_then(|x| x.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        if email.is_empty() || !email.contains('@') {
            return Err("fake-email-site: invalid new email response".into());
        }
        Ok(EmailInfo {
            channel: Channel::FakeEmailSite,
            email,
            token: None,
            expires_at: None,
            created_at: None,
        })
    })
}

/// 获取邮件列表
pub fn get_emails(email: &str) -> Result<Vec<Email>, String> {
    let em = email.trim();
    if em.is_empty() {
        return Err("fake-email-site: empty email".into());
    }
    block_on(async {
        let url = format!(
            "{}/inbox/poll?address={}",
            API_BASE,
            urlencoding::encode(em)
        );
        let resp = headers(http_client().get(url))
            .send()
            .await
            .map_err(|e| e.to_string())?;
        if resp.status().as_u16() == 404 {
            return Ok(Vec::new());
        }
        if !resp.status().is_success() {
            return Err(format!("fake-email-site poll {}", resp.status()));
        }
        let data: Value = resp.json().await.map_err(|e| e.to_string())?;
        let rows = data
            .get("messages")
            .and_then(|x| x.as_array())
            .cloned()
            .unwrap_or_default();
        let mut out = Vec::new();
        for msg in rows {
            let normalized = json!({
                "id": msg.get("id").cloned().unwrap_or(Value::String(String::new())),
                "from": msg.get("from_addr").cloned().unwrap_or(Value::String(String::new())),
                "to": msg.get("to_addr").cloned().unwrap_or(Value::String(String::new())),
                "subject": msg.get("subject").cloned().unwrap_or(Value::String(String::new())),
                "text": msg.get("body_text").cloned().unwrap_or(Value::String(String::new())),
                "timestamp": msg.get("received_at").cloned().unwrap_or(Value::Null),
                "read": Value::Bool(false),
            });
            out.push(normalize_email(&normalized, em));
        }
        Ok(out)
    })
}
